// @cloudscape-design/components/internal/components/options-list/utils/use-highlight-option.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { useCallback, useState } from 'react';
export function createHighlightedOptionHook(_a) {
    var isHighlightable = _a.isHighlightable;
    return function useHighlightedOption(options) {
        var _a = useState(-1), highlightedIndex = _a[0], setHighlightedIndex = _a[1];
        var highlightedOption = options[highlightedIndex] && isHighlightable(options[highlightedIndex]) ? options[highlightedIndex] : undefined;
        var moveHighlightFrom = function (direction, startIndex) {
            if (startIndex === void 0) { startIndex = highlightedIndex; }
            var newIndex = startIndex;
            do {
                newIndex += direction;
            } while (options[newIndex] && !isHighlightable(options[newIndex]));
            if (options[newIndex]) {
                setHighlightedIndex(newIndex);
            }
        };
        var moveHighlight = function (direction) { return moveHighlightFrom(direction); };
        var highlightOption = useCallback(function (option) {
            var index = options.indexOf(option);
            setHighlightedIndex(index);
        }, [options]);
        return {
            setHighlightedIndex: setHighlightedIndex,
            highlightedIndex: highlightedIndex,
            highlightedOption: highlightedOption,
            moveHighlight: moveHighlight,
            resetHighlight: function () { return setHighlightedIndex(-1); },
            goHome: function () { return moveHighlightFrom(1, -1); },
            goEnd: function () { return moveHighlightFrom(-1, options.length); },
            highlightOption: highlightOption
        };
    };
}
