// @cloudscape-design/collection-hooks/dist/cjs/use-collection-state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useCollectionState = void 0;
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
var react_1 = require("react");
var utils_js_1 = require("./utils.js");
function useCollectionState(options, collectionRef) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j;
    var _k = (0, react_1.useReducer)(utils_js_1.collectionReducer, {
        selectedItems: (_b = (_a = options.selection) === null || _a === void 0 ? void 0 : _a.defaultSelectedItems) !== null && _b !== void 0 ? _b : [],
        sortingState: (_c = options.sorting) === null || _c === void 0 ? void 0 : _c.defaultState,
        currentPageIndex: (_e = (_d = options.pagination) === null || _d === void 0 ? void 0 : _d.defaultPage) !== null && _e !== void 0 ? _e : 1,
        filteringText: (_g = (_f = options.filtering) === null || _f === void 0 ? void 0 : _f.defaultFilteringText) !== null && _g !== void 0 ? _g : '',
        propertyFilteringQuery: (_j = (_h = options.propertyFiltering) === null || _h === void 0 ? void 0 : _h.defaultQuery) !== null && _j !== void 0 ? _j : { tokens: [], operation: 'and' },
    }), state = _k[0], dispatch = _k[1];
    return [
        state,
        (0, utils_js_1.createActions)({
            dispatch: dispatch,
            collectionRef: collectionRef,
        }),
    ];
}
exports.useCollectionState = useCollectionState;
