// @cloudscape-design/components/property-filter/property-filter-autosuggest.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { __assign, __rest } from "tslib";
import React, { useRef } from 'react';
import { useAutosuggestItems } from '../autosuggest/options-controller';
import { useDropdownStatus } from '../internal/components/dropdown-status';
import DropdownFooter from '../internal/components/dropdown-footer';
import { generateUniqueId, useUniqueId } from '../internal/hooks/use-unique-id';
import { fireNonCancelableEvent, } from '../internal/events';
import styles from '../autosuggest/styles.css.js';
import { fireCancelableEvent } from '../internal/events/index';
import AutosuggestOptionsList from '../autosuggest/options-list';
import { useAutosuggestLoadMore } from '../autosuggest/load-more-controller';
import AutosuggestInput from '../internal/components/autosuggest-input';
import { useMergeRefs } from '../internal/hooks/use-merge-refs';
var DROPDOWN_WIDTH = 300;
var PropertyFilterAutosuggest = React.forwardRef(function (props, ref) {
    var _a;
    var value = props.value, onChange = props.onChange, onFocus = props.onFocus, onBlur = props.onBlur, onLoadItems = props.onLoadItems, options = props.options, _b = props.filteringType, filteringType = _b === void 0 ? 'auto' : _b, _c = props.statusType, statusType = _c === void 0 ? 'finished' : _c, placeholder = props.placeholder, disabled = props.disabled, ariaLabel = props.ariaLabel, enteredTextLabel = props.enteredTextLabel, onKeyDown = props.onKeyDown, virtualScroll = props.virtualScroll, expandToViewport = props.expandToViewport, filterText = props.filterText, onOptionClick = props.onOptionClick, hideEnteredTextOption = props.hideEnteredTextOption, rest = __rest(props, ["value", "onChange", "onFocus", "onBlur", "onLoadItems", "options", "filteringType", "statusType", "placeholder", "disabled", "ariaLabel", "enteredTextLabel", "onKeyDown", "virtualScroll", "expandToViewport", "filterText", "onOptionClick", "hideEnteredTextOption"]);
    var highlightText = filterText === undefined ? value : filterText;
    var autosuggestInputRef = useRef(null);
    var mergedRef = useMergeRefs(autosuggestInputRef, ref);
    var _d = useAutosuggestItems({
        options: options || [],
        filterValue: value,
        filterText: highlightText,
        filteringType: filteringType,
        hideEnteredTextLabel: hideEnteredTextOption,
        onSelectItem: function (option) {
            var _a;
            var value = option.value || '';
            fireNonCancelableEvent(onChange, { value: value });
            var selectedCancelled = fireCancelableEvent(onOptionClick, option);
            if (!selectedCancelled) {
                (_a = autosuggestInputRef.current) === null || _a === void 0 ? void 0 : _a.close();
            }
            else {
                autosuggestItemsHandlers.resetHighlightWithKeyboard();
            }
        }
    }), autosuggestItemsState = _d[0], autosuggestItemsHandlers = _d[1];
    var autosuggestLoadMoreHandlers = useAutosuggestLoadMore({
        options: options,
        statusType: statusType,
        onLoadItems: function (detail) { return fireNonCancelableEvent(onLoadItems, detail); }
    });
    var handleChange = function (event) {
        autosuggestItemsHandlers.resetHighlightWithKeyboard();
        fireNonCancelableEvent(onChange, event.detail);
    };
    var handleDelayedInput = function (event) {
        autosuggestLoadMoreHandlers.fireLoadMoreOnInputChange(event.detail.value);
    };
    var handleFocus = function () {
        autosuggestLoadMoreHandlers.fireLoadMoreOnInputFocus();
        fireCancelableEvent(onFocus, null);
    };
    var handleBlur = function () {
        fireCancelableEvent(onBlur, null);
    };
    var handleKeyDown = function (e) {
        fireCancelableEvent(onKeyDown, e.detail);
    };
    var handlePressArrowDown = function () {
        autosuggestItemsHandlers.moveHighlightWithKeyboard(1);
    };
    var handlePressArrowUp = function () {
        autosuggestItemsHandlers.moveHighlightWithKeyboard(-1);
    };
    var handlePressEnter = function () {
        return autosuggestItemsHandlers.selectHighlightedOptionWithKeyboard();
    };
    var handleCloseDropdown = function () {
        autosuggestItemsHandlers.resetHighlightWithKeyboard();
    };
    var handleRecoveryClick = function () {
        var _a;
        autosuggestLoadMoreHandlers.fireLoadMoreOnRecoveryClick();
        (_a = autosuggestInputRef.current) === null || _a === void 0 ? void 0 : _a.focus();
    };
    var handleDropdownMouseDown = function (event) {
        // Prevent currently focused element from losing focus.
        event.preventDefault();
    };
    var selfControlId = useUniqueId('input');
    var controlId = (_a = rest.controlId) !== null && _a !== void 0 ? _a : selfControlId;
    var listId = useUniqueId('list');
    var highlightedOptionId = autosuggestItemsState.highlightedOption ? generateUniqueId() : undefined;
    var isEmpty = !value && !autosuggestItemsState.items.length;
    var dropdownStatus = useDropdownStatus(__assign(__assign({}, props), { isEmpty: isEmpty, onRecoveryClick: handleRecoveryClick }));
    return (React.createElement(AutosuggestInput, __assign({ ref: mergedRef, className: styles.root, value: value, onChange: handleChange, onFocus: handleFocus, onBlur: handleBlur, onKeyDown: handleKeyDown, controlId: controlId, placeholder: placeholder, disabled: disabled, ariaLabel: ariaLabel, expandToViewport: expandToViewport, ariaControls: listId, ariaActivedescendant: highlightedOptionId, dropdownExpanded: autosuggestItemsState.items.length > 1, dropdownContent: React.createElement(AutosuggestOptionsList, { autosuggestItemsState: autosuggestItemsState, autosuggestItemsHandlers: autosuggestItemsHandlers, highlightedOptionId: highlightedOptionId, highlightText: highlightText, listId: listId, controlId: controlId, enteredTextLabel: enteredTextLabel, handleLoadMore: autosuggestLoadMoreHandlers.fireLoadMoreOnScroll, hasDropdownStatus: dropdownStatus.content !== null, virtualScroll: virtualScroll, listBottom: !dropdownStatus.isSticky ? React.createElement(DropdownFooter, { content: dropdownStatus.content }) : null }), dropdownFooter: dropdownStatus.isSticky ? (React.createElement(DropdownFooter, { content: dropdownStatus.content, hasItems: autosuggestItemsState.items.length >= 1 })) : null, dropdownWidth: DROPDOWN_WIDTH, onDropdownMouseDown: handleDropdownMouseDown, onCloseDropdown: handleCloseDropdown, onDelayedInput: handleDelayedInput, onPressArrowDown: handlePressArrowDown, onPressArrowUp: handlePressArrowUp, onPressEnter: handlePressEnter }, rest)));
});
export default PropertyFilterAutosuggest;
