// @cloudscape-design/components/test-utils/selectors/date-picker/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CalendarWrapper = void 0;
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
var selectors_1 = require("@cloudscape-design/test-utils-core/selectors");
var styles_selectors_js_1 = require("../../../date-picker/styles.selectors.js");
var button_1 = require("../button");
var base_input_1 = require("../input/base-input");
var dropdown_1 = require("../internal/dropdown");
var DatePickerWrapper = /** @class */ (function (_super) {
    __extends(DatePickerWrapper, _super);
    function DatePickerWrapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @param options
     * * expandToViewport (boolean) - Use this when the component under test is rendered with an `expandToViewport` flag.
     */
    DatePickerWrapper.prototype.findCalendar = function (options) {
        if (options === void 0) { options = {
            expandToViewport: false
        }; }
        var wrapper = options.expandToViewport ? (0, selectors_1.createWrapper)() : this;
        return wrapper.findComponent(".".concat(styles_selectors_js_1.default.calendar), CalendarWrapper);
    };
    DatePickerWrapper.prototype.findCalendarDropdown = function () {
        var dropdown = new dropdown_1.default(this.getElement());
        return dropdown.findOpenDropdown();
    };
    DatePickerWrapper.prototype.findOpenCalendarButton = function () {
        return this.findComponent(".".concat(styles_selectors_js_1.default['open-calendar-button']), button_1.default);
    };
    DatePickerWrapper.rootSelector = styles_selectors_js_1.default.root;
    return DatePickerWrapper;
}(base_input_1.default));
exports.default = DatePickerWrapper;
var CalendarWrapper = /** @class */ (function (_super) {
    __extends(CalendarWrapper, _super);
    function CalendarWrapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Returns a day container on the calendar.
     *
     * @param row 1-based row index of the day.
     * @param column 1-based column index of the day.
     */
    CalendarWrapper.prototype.findDateAt = function (row, column) {
        return this.find(".".concat(styles_selectors_js_1.default['calendar-week'], ":nth-child(").concat(row, ") .").concat(styles_selectors_js_1.default['calendar-day'], ":nth-child(").concat(column, ")"));
    };
    CalendarWrapper.prototype.findHeader = function () {
        return this.findByClassName(styles_selectors_js_1.default['calendar-header']);
    };
    CalendarWrapper.prototype.findPreviousMonthButton = function () {
        return this.findComponent(".".concat(styles_selectors_js_1.default['calendar-prev-month-btn']), button_1.default);
    };
    CalendarWrapper.prototype.findNextMonthButton = function () {
        return this.findComponent(".".concat(styles_selectors_js_1.default['calendar-next-month-btn']), button_1.default);
    };
    CalendarWrapper.prototype.findSelectedDate = function () {
        return this.find(".".concat(styles_selectors_js_1.default['calendar-day-selected']));
    };
    return CalendarWrapper;
}(selectors_1.ComponentWrapper));
exports.CalendarWrapper = CalendarWrapper;
