// @opentelemetry/instrumentation-mysql/build/src/instrumentation.js
"use strict";
/*
 * Copyright The OpenTelemetry Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MySQLInstrumentation = void 0;
const api_1 = require("@opentelemetry/api");
const instrumentation_1 = require("@opentelemetry/instrumentation");
const semantic_conventions_1 = require("@opentelemetry/semantic-conventions");
const utils_1 = require("./utils");
const version_1 = require("./version");
class MySQLInstrumentation extends instrumentation_1.InstrumentationBase {
    constructor(config) {
        super('@opentelemetry/instrumentation-mysql', version_1.VERSION, config);
    }
    init() {
        return [
            new instrumentation_1.InstrumentationNodeModuleDefinition('mysql', ['2.*'], (moduleExports, moduleVersion) => {
                api_1.diag.debug(`Patching mysql@${moduleVersion}`);
                api_1.diag.debug('Patching mysql.createConnection');
                if (instrumentation_1.isWrapped(moduleExports.createConnection)) {
                    this._unwrap(moduleExports, 'createConnection');
                }
                this._wrap(moduleExports, 'createConnection', this._patchCreateConnection(moduleExports.format));
                api_1.diag.debug('Patching mysql.createPool');
                if (instrumentation_1.isWrapped(moduleExports.createPool)) {
                    this._unwrap(moduleExports, 'createPool');
                }
                this._wrap(moduleExports, 'createPool', this._patchCreatePool(moduleExports.format));
                api_1.diag.debug('Patching mysql.createPoolCluster');
                if (instrumentation_1.isWrapped(moduleExports.createPoolCluster)) {
                    this._unwrap(moduleExports, 'createPoolCluster');
                }
                this._wrap(moduleExports, 'createPoolCluster', this._patchCreatePoolCluster(moduleExports.format));
                return moduleExports;
            }, moduleExports => {
                if (moduleExports === undefined)
                    return;
                this._unwrap(moduleExports, 'createConnection');
                this._unwrap(moduleExports, 'createPool');
                this._unwrap(moduleExports, 'createPoolCluster');
            }),
        ];
    }
    // global export function
    _patchCreateConnection(format) {
        return (originalCreateConnection) => {
            const thisPlugin = this;
            api_1.diag.debug('MySQLInstrumentation#patch: patched mysql createConnection');
            return function createConnection(_connectionUri) {
                const originalResult = originalCreateConnection(...arguments);
                // This is unwrapped on next call after unpatch
                thisPlugin._wrap(originalResult, 'query', thisPlugin._patchQuery(originalResult, format));
                return originalResult;
            };
        };
    }
    // global export function
    _patchCreatePool(format) {
        return (originalCreatePool) => {
            const thisPlugin = this;
            api_1.diag.debug('MySQLInstrumentation#patch: patched mysql createPool');
            return function createPool(_config) {
                const pool = originalCreatePool(...arguments);
                thisPlugin._wrap(pool, 'query', thisPlugin._patchQuery(pool, format));
                thisPlugin._wrap(pool, 'getConnection', thisPlugin._patchGetConnection(pool, format));
                return pool;
            };
        };
    }
    // global export function
    _patchCreatePoolCluster(format) {
        return (originalCreatePoolCluster) => {
            const thisPlugin = this;
            api_1.diag.debug('MySQLInstrumentation#patch: patched mysql createPoolCluster');
            return function createPool(_config) {
                const cluster = originalCreatePoolCluster(...arguments);
                // This is unwrapped on next call after unpatch
                thisPlugin._wrap(cluster, 'getConnection', thisPlugin._patchGetConnection(cluster, format));
                return cluster;
            };
        };
    }
    // method on cluster or pool
    _patchGetConnection(pool, format) {
        return (originalGetConnection) => {
            const thisPlugin = this;
            api_1.diag.debug('MySQLInstrumentation#patch: patched mysql pool getConnection');
            return function getConnection(arg1, arg2, arg3) {
                // Unwrap if unpatch has been called
                if (!thisPlugin['_enabled']) {
                    thisPlugin._unwrap(pool, 'getConnection');
                    return originalGetConnection.apply(pool, arguments);
                }
                if (arguments.length === 1 && typeof arg1 === 'function') {
                    const patchFn = thisPlugin._getConnectionCallbackPatchFn(arg1, format);
                    return originalGetConnection.call(pool, patchFn);
                }
                if (arguments.length === 2 && typeof arg2 === 'function') {
                    const patchFn = thisPlugin._getConnectionCallbackPatchFn(arg2, format);
                    return originalGetConnection.call(pool, arg1, patchFn);
                }
                if (arguments.length === 3 && typeof arg3 === 'function') {
                    const patchFn = thisPlugin._getConnectionCallbackPatchFn(arg3, format);
                    return originalGetConnection.call(pool, arg1, arg2, patchFn);
                }
                return originalGetConnection.apply(pool, arguments);
            };
        };
    }
    _getConnectionCallbackPatchFn(cb, format) {
        const thisPlugin = this;
        const activeContext = api_1.context.active();
        return function (err, connection) {
            if (connection) {
                // this is the callback passed into a query
                // no need to unwrap
                if (!instrumentation_1.isWrapped(connection.query)) {
                    thisPlugin._wrap(connection, 'query', thisPlugin._patchQuery(connection, format));
                }
            }
            if (typeof cb === 'function') {
                api_1.context.with(activeContext, cb, this, err, connection);
            }
        };
    }
    _patchQuery(connection, format) {
        return (originalQuery) => {
            const thisPlugin = this;
            api_1.diag.debug('MySQLInstrumentation: patched mysql query');
            return function query(query, _valuesOrCallback, _callback) {
                if (!thisPlugin['_enabled']) {
                    thisPlugin._unwrap(connection, 'query');
                    return originalQuery.apply(connection, arguments);
                }
                const span = thisPlugin.tracer.startSpan(utils_1.getSpanName(query), {
                    kind: api_1.SpanKind.CLIENT,
                    attributes: Object.assign(Object.assign({}, MySQLInstrumentation.COMMON_ATTRIBUTES), utils_1.getConnectionAttributes(connection.config)),
                });
                let values;
                if (Array.isArray(_valuesOrCallback)) {
                    values = _valuesOrCallback;
                }
                else if (arguments[2]) {
                    values = [_valuesOrCallback];
                }
                span.setAttribute(semantic_conventions_1.SemanticAttributes.DB_STATEMENT, utils_1.getDbStatement(query, format, values));
                const cbIndex = Array.from(arguments).findIndex(arg => typeof arg === 'function');
                if (cbIndex === -1) {
                    const streamableQuery = originalQuery.apply(connection, arguments);
                    return streamableQuery
                        .on('error', err => span.setStatus({
                        code: api_1.SpanStatusCode.ERROR,
                        message: err.message,
                    }))
                        .on('end', () => {
                        span.end();
                    });
                }
                else {
                    thisPlugin._wrap(arguments, cbIndex, thisPlugin._patchCallbackQuery(span));
                    return originalQuery.apply(connection, arguments);
                }
            };
        };
    }
    _patchCallbackQuery(span) {
        return (originalCallback) => {
            return function (err, results, fields) {
                if (err) {
                    span.setStatus({
                        code: api_1.SpanStatusCode.ERROR,
                        message: err.message,
                    });
                }
                span.end();
                return originalCallback(...arguments);
            };
        };
    }
}
exports.MySQLInstrumentation = MySQLInstrumentation;
MySQLInstrumentation.COMMON_ATTRIBUTES = {
    [semantic_conventions_1.SemanticAttributes.DB_SYSTEM]: semantic_conventions_1.DbSystemValues.MYSQL,
};
