// @opentelemetry/instrumentation-aws-sdk/build/src/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bindPromise = exports.extractAttributesFromNormalizedRequest = exports.normalizeV3Request = exports.normalizeV2Request = exports.removeSuffixFromStringIfExists = void 0;
const api_1 = require("@opentelemetry/api");
const semantic_conventions_1 = require("@opentelemetry/semantic-conventions");
const enums_1 = require("./enums");
const toPascalCase = (str) => typeof str === 'string' ? str.charAt(0).toUpperCase() + str.slice(1) : str;
const removeSuffixFromStringIfExists = (str, suffixToRemove) => {
    const suffixLength = suffixToRemove.length;
    return (str === null || str === void 0 ? void 0 : str.slice(-suffixLength)) === suffixToRemove
        ? str.slice(0, str.length - suffixLength)
        : str;
};
exports.removeSuffixFromStringIfExists = removeSuffixFromStringIfExists;
const normalizeV2Request = (awsV2Request) => {
    var _a, _b, _c, _d, _e;
    const service = (_a = awsV2Request) === null || _a === void 0 ? void 0 : _a.service;
    return {
        serviceName: (_c = (_b = service === null || service === void 0 ? void 0 : service.api) === null || _b === void 0 ? void 0 : _b.serviceId) === null || _c === void 0 ? void 0 : _c.replace(/\s+/g, ''),
        commandName: toPascalCase((_d = awsV2Request) === null || _d === void 0 ? void 0 : _d.operation),
        commandInput: awsV2Request.params,
        region: (_e = service === null || service === void 0 ? void 0 : service.config) === null || _e === void 0 ? void 0 : _e.region,
    };
};
exports.normalizeV2Request = normalizeV2Request;
const normalizeV3Request = (serviceName, commandNameWithSuffix, commandInput, region) => {
    return {
        serviceName: serviceName === null || serviceName === void 0 ? void 0 : serviceName.replace(/\s+/g, ''),
        commandName: exports.removeSuffixFromStringIfExists(commandNameWithSuffix, 'Command'),
        commandInput,
        region,
    };
};
exports.normalizeV3Request = normalizeV3Request;
const extractAttributesFromNormalizedRequest = (normalizedRequest) => {
    return {
        [semantic_conventions_1.SemanticAttributes.RPC_SYSTEM]: 'aws-api',
        [semantic_conventions_1.SemanticAttributes.RPC_METHOD]: normalizedRequest.commandName,
        [semantic_conventions_1.SemanticAttributes.RPC_SERVICE]: normalizedRequest.serviceName,
        [enums_1.AttributeNames.AWS_REGION]: normalizedRequest.region,
    };
};
exports.extractAttributesFromNormalizedRequest = extractAttributesFromNormalizedRequest;
const bindPromise = (target, contextForCallbacks, rebindCount = 1) => {
    const origThen = target.then;
    target.then = function (onFulfilled, onRejected) {
        const newOnFulfilled = api_1.context.bind(contextForCallbacks, onFulfilled);
        const newOnRejected = api_1.context.bind(contextForCallbacks, onRejected);
        const patchedPromise = origThen.call(this, newOnFulfilled, newOnRejected);
        return rebindCount > 1
            ? exports.bindPromise(patchedPromise, contextForCallbacks, rebindCount - 1)
            : patchedPromise;
    };
    return target;
};
exports.bindPromise = bindPromise;
