// @cloudscape-design/components/internal/generated/styles/tokens.js
export var colorChartsRed300 = "var(--color-charts-red-300-taex3d, #ea7158)";
export var colorChartsRed400 = "var(--color-charts-red-400-xo1w45, #dc5032)";
export var colorChartsRed500 = "var(--color-charts-red-500-8et8sv, #d13313)";
export var colorChartsRed600 = "var(--color-charts-red-600-ej8iox, #ba2e0f)";
export var colorChartsRed700 = "var(--color-charts-red-700-pe7xt7, #a82a0c)";
export var colorChartsRed800 = "var(--color-charts-red-800-9wqz57, #972709)";
export var colorChartsRed900 = "var(--color-charts-red-900-fkiox3, #892407)";
export var colorChartsRed1000 = "var(--color-charts-red-1000-xhb2ma, #7d2105)";
export var colorChartsRed1100 = "var(--color-charts-red-1100-zkhjl8, #721e03)";
export var colorChartsRed1200 = "var(--color-charts-red-1200-t42yl7, #671c00)";
export var colorChartsOrange300 = "var(--color-charts-orange-300-uh1x0z, #e07941)";
export var colorChartsOrange400 = "var(--color-charts-orange-400-rg9ft0, #cc5f21)";
export var colorChartsOrange500 = "var(--color-charts-orange-500-6dcz8x, #bc4d01)";
export var colorChartsOrange600 = "var(--color-charts-orange-600-psg5i7, #a84401)";
export var colorChartsOrange700 = "var(--color-charts-orange-700-6rlkwo, #983c02)";
export var colorChartsOrange800 = "var(--color-charts-orange-800-aqh9u5, #8a3603)";
export var colorChartsOrange900 = "var(--color-charts-orange-900-71d364, #7e3103)";
export var colorChartsOrange1000 = "var(--color-charts-orange-1000-p3xy9d, #732c02)";
export var colorChartsOrange1100 = "var(--color-charts-orange-1100-plp0ix, #692801)";
export var colorChartsOrange1200 = "var(--color-charts-orange-1200-msf9v8, #602400)";
export var colorChartsYellow300 = "var(--color-charts-yellow-300-zgwohc, #b2911c)";
export var colorChartsYellow400 = "var(--color-charts-yellow-400-z6s56e, #9c7b0b)";
export var colorChartsYellow500 = "var(--color-charts-yellow-500-f52bpo, #8a6b05)";
export var colorChartsYellow600 = "var(--color-charts-yellow-600-gzedvz, #7b5f04)";
export var colorChartsYellow700 = "var(--color-charts-yellow-700-uq1jq9, #6f5504)";
export var colorChartsYellow800 = "var(--color-charts-yellow-800-esbvam, #654d03)";
export var colorChartsYellow900 = "var(--color-charts-yellow-900-e2ugyo, #5d4503)";
export var colorChartsYellow1000 = "var(--color-charts-yellow-1000-myf1mq, #553f03)";
export var colorChartsYellow1100 = "var(--color-charts-yellow-1100-fi5c3g, #4d3901)";
export var colorChartsYellow1200 = "var(--color-charts-yellow-1200-faeu24, #483300)";
export var colorChartsGreen300 = "var(--color-charts-green-300-8q7o65, #67a353)";
export var colorChartsGreen400 = "var(--color-charts-green-400-b5efwo, #41902c)";
export var colorChartsGreen500 = "var(--color-charts-green-500-o7urau, #1f8104)";
export var colorChartsGreen600 = "var(--color-charts-green-600-8e3las, #1a7302)";
export var colorChartsGreen700 = "var(--color-charts-green-700-h4agmh, #176702)";
export var colorChartsGreen800 = "var(--color-charts-green-800-jp1xwf, #145d02)";
export var colorChartsGreen900 = "var(--color-charts-green-900-czmqk3, #125502)";
export var colorChartsGreen1000 = "var(--color-charts-green-1000-wrs7fm, #104d01)";
export var colorChartsGreen1100 = "var(--color-charts-green-1100-d8mgjl, #0f4601)";
export var colorChartsGreen1200 = "var(--color-charts-green-1200-jyfc23, #0d4000)";
export var colorChartsTeal300 = "var(--color-charts-teal-300-egfaot, #2ea597)";
export var colorChartsTeal400 = "var(--color-charts-teal-400-ikctdb, #1c8e81)";
export var colorChartsTeal500 = "var(--color-charts-teal-500-v952sw, #0d7d70)";
export var colorChartsTeal600 = "var(--color-charts-teal-600-9ffc3q, #096f64)";
export var colorChartsTeal700 = "var(--color-charts-teal-700-nn7a5u, #06645a)";
export var colorChartsTeal800 = "var(--color-charts-teal-800-97l4n7, #045b52)";
export var colorChartsTeal900 = "var(--color-charts-teal-900-zq1u0w, #03524a)";
export var colorChartsTeal1000 = "var(--color-charts-teal-1000-u4e4vv, #014b44)";
export var colorChartsTeal1100 = "var(--color-charts-teal-1100-cdt1d8, #01443e)";
export var colorChartsTeal1200 = "var(--color-charts-teal-1200-3ozaok, #003e38)";
export var colorChartsBlue1300 = "var(--color-charts-blue-1-300-4x8rnx, #529ccb)";
export var colorChartsBlue1400 = "var(--color-charts-blue-1-400-owg3dt, #3184c2)";
export var colorChartsBlue1500 = "var(--color-charts-blue-1-500-sl8chl, #0273bb)";
export var colorChartsBlue1600 = "var(--color-charts-blue-1-600-3824a7, #0166ab)";
export var colorChartsBlue1700 = "var(--color-charts-blue-1-700-e3d0uw, #015b9d)";
export var colorChartsBlue1800 = "var(--color-charts-blue-1-800-vnypf9, #015292)";
export var colorChartsBlue1900 = "var(--color-charts-blue-1-900-rs4k5z, #014a87)";
export var colorChartsBlue11000 = "var(--color-charts-blue-1-1000-wjr7hn, #01437d)";
export var colorChartsBlue11100 = "var(--color-charts-blue-1-1100-nmn22x, #003c75)";
export var colorChartsBlue11200 = "var(--color-charts-blue-1-1200-od4mlj, #00366d)";
export var colorChartsBlue2300 = "var(--color-charts-blue-2-300-b0dh81, #7492e7)";
export var colorChartsBlue2400 = "var(--color-charts-blue-2-400-iba8cu, #5978e3)";
export var colorChartsBlue2500 = "var(--color-charts-blue-2-500-vtdvkt, #4066df)";
export var colorChartsBlue2600 = "var(--color-charts-blue-2-600-tpo4rj, #3759ce)";
export var colorChartsBlue2700 = "var(--color-charts-blue-2-700-p6tu56, #314fbf)";
export var colorChartsBlue2800 = "var(--color-charts-blue-2-800-08f3pp, #2c46b1)";
export var colorChartsBlue2900 = "var(--color-charts-blue-2-900-3aq93e, #273ea5)";
export var colorChartsBlue21000 = "var(--color-charts-blue-2-1000-6b05hm, #23379b)";
export var colorChartsBlue21100 = "var(--color-charts-blue-2-1100-0014sa, #1f3191)";
export var colorChartsBlue21200 = "var(--color-charts-blue-2-1200-tzr398, #1b2b88)";
export var colorChartsPurple300 = "var(--color-charts-purple-300-yatxj8, #a783e1)";
export var colorChartsPurple400 = "var(--color-charts-purple-400-d384pt, #9469d6)";
export var colorChartsPurple500 = "var(--color-charts-purple-500-skntta, #8456ce)";
export var colorChartsPurple600 = "var(--color-charts-purple-600-8iqvzu, #7749bf)";
export var colorChartsPurple700 = "var(--color-charts-purple-700-3qzy9s, #6b40b2)";
export var colorChartsPurple800 = "var(--color-charts-purple-800-ucyxda, #6237a7)";
export var colorChartsPurple900 = "var(--color-charts-purple-900-tlepdb, #59309d)";
export var colorChartsPurple1000 = "var(--color-charts-purple-1000-3pjp6l, #512994)";
export var colorChartsPurple1100 = "var(--color-charts-purple-1100-t0m9qa, #4a238b)";
export var colorChartsPurple1200 = "var(--color-charts-purple-1200-003tqq, #431d84)";
export var colorChartsPink300 = "var(--color-charts-pink-300-0gug9u, #da7596)";
export var colorChartsPink400 = "var(--color-charts-pink-400-jh84a9, #ce567c)";
export var colorChartsPink500 = "var(--color-charts-pink-500-c97yed, #c33d69)";
export var colorChartsPink600 = "var(--color-charts-pink-600-rpo3pi, #b1325c)";
export var colorChartsPink700 = "var(--color-charts-pink-700-mshe2z, #a32952)";
export var colorChartsPink800 = "var(--color-charts-pink-800-bph464, #962249)";
export var colorChartsPink900 = "var(--color-charts-pink-900-mvfm0y, #8b1b42)";
export var colorChartsPink1000 = "var(--color-charts-pink-1000-kq8pu0, #81143b)";
export var colorChartsPink1100 = "var(--color-charts-pink-1100-pi0m3u, #780d35)";
export var colorChartsPink1200 = "var(--color-charts-pink-1200-hukz4p, #6f062f)";
export var colorChartsStatusCritical = "var(--color-charts-status-critical-268let, #7d2105)";
export var colorChartsStatusHigh = "var(--color-charts-status-high-vsd25n, #ba2e0f)";
export var colorChartsStatusMedium = "var(--color-charts-status-medium-371m8g, #cc5f21)";
export var colorChartsStatusLow = "var(--color-charts-status-low-lfwc29, #b2911c)";
export var colorChartsStatusPositive = "var(--color-charts-status-positive-2ozsqj, #67a353)";
export var colorChartsStatusInfo = "var(--color-charts-status-info-l8g84h, #3184c2)";
export var colorChartsStatusNeutral = "var(--color-charts-status-neutral-b4b0p2, #7d8998)";
export var colorChartsThresholdNegative = "var(--color-charts-threshold-negative-y7lrhf, #d91515)";
export var colorChartsThresholdPositive = "var(--color-charts-threshold-positive-42ybvt, #037f0c)";
export var colorChartsThresholdInfo = "var(--color-charts-threshold-info-kcjduw, #0972d3)";
export var colorChartsThresholdNeutral = "var(--color-charts-threshold-neutral-k14cuc, #5f6b7a)";
export var colorChartsPaletteCategorical1 = "var(--color-charts-palette-categorical-1-ivmh5s, #7492e7)";
export var colorChartsPaletteCategorical2 = "var(--color-charts-palette-categorical-2-emtquh, #c33d69)";
export var colorChartsPaletteCategorical3 = "var(--color-charts-palette-categorical-3-imrqvq, #2ea597)";
export var colorChartsPaletteCategorical4 = "var(--color-charts-palette-categorical-4-g31h69, #8456ce)";
export var colorChartsPaletteCategorical5 = "var(--color-charts-palette-categorical-5-64f3p2, #e07941)";
export var colorChartsPaletteCategorical6 = "var(--color-charts-palette-categorical-6-msr86m, #3759ce)";
export var colorChartsPaletteCategorical7 = "var(--color-charts-palette-categorical-7-kdjms7, #962249)";
export var colorChartsPaletteCategorical8 = "var(--color-charts-palette-categorical-8-79776a, #096f64)";
export var colorChartsPaletteCategorical9 = "var(--color-charts-palette-categorical-9-s4w3dd, #6237a7)";
export var colorChartsPaletteCategorical10 = "var(--color-charts-palette-categorical-10-cuicpx, #a84401)";
export var colorChartsPaletteCategorical11 = "var(--color-charts-palette-categorical-11-no2uu5, #273ea5)";
export var colorChartsPaletteCategorical12 = "var(--color-charts-palette-categorical-12-jq9rcw, #780d35)";
export var colorChartsPaletteCategorical13 = "var(--color-charts-palette-categorical-13-4cawsz, #03524a)";
export var colorChartsPaletteCategorical14 = "var(--color-charts-palette-categorical-14-i1n5cl, #4a238b)";
export var colorChartsPaletteCategorical15 = "var(--color-charts-palette-categorical-15-9pgabq, #7e3103)";
export var colorChartsPaletteCategorical16 = "var(--color-charts-palette-categorical-16-wnvn1q, #1b2b88)";
export var colorChartsPaletteCategorical17 = "var(--color-charts-palette-categorical-17-y2duhb, #ce567c)";
export var colorChartsPaletteCategorical18 = "var(--color-charts-palette-categorical-18-xehfp2, #003e38)";
export var colorChartsPaletteCategorical19 = "var(--color-charts-palette-categorical-19-v1x5n3, #9469d6)";
export var colorChartsPaletteCategorical20 = "var(--color-charts-palette-categorical-20-q5g6qv, #602400)";
export var colorChartsPaletteCategorical21 = "var(--color-charts-palette-categorical-21-z2fito, #4066df)";
export var colorChartsPaletteCategorical22 = "var(--color-charts-palette-categorical-22-mdbc68, #a32952)";
export var colorChartsPaletteCategorical23 = "var(--color-charts-palette-categorical-23-9iyat0, #0d7d70)";
export var colorChartsPaletteCategorical24 = "var(--color-charts-palette-categorical-24-17kvjc, #6b40b2)";
export var colorChartsPaletteCategorical25 = "var(--color-charts-palette-categorical-25-c2vakt, #bc4d01)";
export var colorChartsPaletteCategorical26 = "var(--color-charts-palette-categorical-26-g98tv5, #2c46b1)";
export var colorChartsPaletteCategorical27 = "var(--color-charts-palette-categorical-27-lo8vzl, #81143b)";
export var colorChartsPaletteCategorical28 = "var(--color-charts-palette-categorical-28-2ppjnc, #045b52)";
export var colorChartsPaletteCategorical29 = "var(--color-charts-palette-categorical-29-68gzom, #512994)";
export var colorChartsPaletteCategorical30 = "var(--color-charts-palette-categorical-30-9pbejo, #8a3603)";
export var colorChartsPaletteCategorical31 = "var(--color-charts-palette-categorical-31-8odgjs, #1f3191)";
export var colorChartsPaletteCategorical32 = "var(--color-charts-palette-categorical-32-xug28a, #da7596)";
export var colorChartsPaletteCategorical33 = "var(--color-charts-palette-categorical-33-cr6l8a, #01443e)";
export var colorChartsPaletteCategorical34 = "var(--color-charts-palette-categorical-34-209022, #a783e1)";
export var colorChartsPaletteCategorical35 = "var(--color-charts-palette-categorical-35-twgdku, #692801)";
export var colorChartsPaletteCategorical36 = "var(--color-charts-palette-categorical-36-g2mar5, #5978e3)";
export var colorChartsPaletteCategorical37 = "var(--color-charts-palette-categorical-37-9deaq2, #b1325c)";
export var colorChartsPaletteCategorical38 = "var(--color-charts-palette-categorical-38-5fg2jv, #1c8e81)";
export var colorChartsPaletteCategorical39 = "var(--color-charts-palette-categorical-39-c8yeil, #7749bf)";
export var colorChartsPaletteCategorical40 = "var(--color-charts-palette-categorical-40-vlagy9, #cc5f21)";
export var colorChartsPaletteCategorical41 = "var(--color-charts-palette-categorical-41-o49tuv, #314fbf)";
export var colorChartsPaletteCategorical42 = "var(--color-charts-palette-categorical-42-5irquw, #8b1b42)";
export var colorChartsPaletteCategorical43 = "var(--color-charts-palette-categorical-43-8pxo9j, #06645a)";
export var colorChartsPaletteCategorical44 = "var(--color-charts-palette-categorical-44-w3yd1t, #59309d)";
export var colorChartsPaletteCategorical45 = "var(--color-charts-palette-categorical-45-32a8x6, #983c02)";
export var colorChartsPaletteCategorical46 = "var(--color-charts-palette-categorical-46-4ikr9l, #23379b)";
export var colorChartsPaletteCategorical47 = "var(--color-charts-palette-categorical-47-wg3s7g, #6f062f)";
export var colorChartsPaletteCategorical48 = "var(--color-charts-palette-categorical-48-7no3ly, #014b44)";
export var colorChartsPaletteCategorical49 = "var(--color-charts-palette-categorical-49-qy0tr5, #431d84)";
export var colorChartsPaletteCategorical50 = "var(--color-charts-palette-categorical-50-gkixcb, #732c02)";
export var colorBackgroundButtonPrimaryActive = "var(--color-background-button-primary-active-vg0knj, #033160)";
export var colorBackgroundButtonPrimaryDefault = "var(--color-background-button-primary-default-pwn1g1, #0972d3)";
export var colorBackgroundButtonPrimaryDisabled = "var(--color-background-button-primary-disabled-eon7km, #e9ebed)";
export var colorBackgroundButtonPrimaryHover = "var(--color-background-button-primary-hover-qu1zio, #033160)";
export var colorBackgroundContainerContent = "var(--color-background-container-content-i8i4a0, #ffffff)";
export var colorBackgroundContainerHeader = "var(--color-background-container-header-4flbq5, #ffffff)";
export var colorBackgroundControlChecked = "var(--color-background-control-checked-9admlu, #0972d3)";
export var colorBackgroundControlDefault = "var(--color-background-control-default-4pa05r, #ffffff)";
export var colorBackgroundControlDisabled = "var(--color-background-control-disabled-zplipy, #d1d5db)";
export var colorBackgroundDropdownItemDefault = "var(--color-background-dropdown-item-default-fzdqsg, #ffffff)";
export var colorBackgroundDropdownItemFilterMatch = "var(--color-background-dropdown-item-filter-match-6ac2mw, #f2f8fd)";
export var colorBackgroundDropdownItemHover = "var(--color-background-dropdown-item-hover-mqsxkv, #f4f4f4)";
export var colorBackgroundHomeHeader = "var(--color-background-home-header-mpq08f, #000716)";
export var colorBackgroundInputDefault = "var(--color-background-input-default-u57378, #ffffff)";
export var colorBackgroundInputDisabled = "var(--color-background-input-disabled-wu9j6u, #e9ebed)";
export var colorBackgroundItemSelected = "var(--color-background-item-selected-ebt4bi, #f2f8fd)";
export var colorBackgroundLayoutMain = "var(--color-background-layout-main-sfhm4y, #ffffff)";
export var colorBackgroundNotificationBlue = "var(--color-background-notification-blue-vqb59v, #0972d3)";
export var colorBackgroundNotificationGreen = "var(--color-background-notification-green-0d9ow8, #037f0c)";
export var colorBackgroundNotificationRed = "var(--color-background-notification-red-v7js1f, #d91515)";
export var colorBackgroundStatusError = "var(--color-background-status-error-pe1hux, #fff7f7)";
export var colorBackgroundStatusInfo = "var(--color-background-status-info-usb30e, #f2f8fd)";
export var colorBackgroundStatusSuccess = "var(--color-background-status-success-xnxxn4, #f2fcf3)";
export var colorBackgroundStatusWarning = "var(--color-background-status-warning-zz4oi6, #ffffff)";
export var colorBackgroundToggleCheckedDisabled = "var(--color-background-toggle-checked-disabled-07lf06, #b5d6f4)";
export var colorBorderButtonNormalDefault = "var(--color-border-button-normal-default-djv8f5, #0972d3)";
export var colorBorderButtonNormalDisabled = "var(--color-border-button-normal-disabled-262a0x, #9ba7b6)";
export var colorBorderButtonNormalHover = "var(--color-border-button-normal-hover-e1d0kh, #033160)";
export var colorBorderButtonPrimaryDisabled = "var(--color-border-button-primary-disabled-6gwabt, #e9ebed)";
export var colorBorderContainerTop = "var(--color-border-container-top-srcvx4, transparent)";
export var colorBorderControlDefault = "var(--color-border-control-default-m3lmsh, #7d8998)";
export var colorBorderDividerDefault = "var(--color-border-divider-default-7s2wjw, #e9ebed)";
export var colorBorderDropdownItemHover = "var(--color-border-dropdown-item-hover-tyzq9m, #7d8998)";
export var colorBorderInputDefault = "var(--color-border-input-default-dtyu8d, #9ba7b6)";
export var colorBorderItemFocused = "var(--color-border-item-focused-ap3b6s, #0972d3)";
export var colorBorderItemSelected = "var(--color-border-item-selected-ppkssz, #0972d3)";
export var colorBorderStatusError = "var(--color-border-status-error-mfrddc, #d91515)";
export var colorBorderStatusInfo = "var(--color-border-status-info-546i7i, #0972d3)";
export var colorBorderStatusSuccess = "var(--color-border-status-success-cqi45f, #037f0c)";
export var colorBorderStatusWarning = "var(--color-border-status-warning-hg85ut, #7d8998)";
export var colorForegroundControlDefault = "var(--color-foreground-control-default-7ajdem, #ffffff)";
export var colorForegroundControlDisabled = "var(--color-foreground-control-disabled-v6a97u, #ffffff)";
export var colorTextAccent = "var(--color-text-accent-s1eqko, #0972d3)";
export var colorTextBodyDefault = "var(--color-text-body-default-ajf1h5, #000716)";
export var colorTextBodySecondary = "var(--color-text-body-secondary-fdstdf, #414d5c)";
export var colorTextButtonNormalActive = "var(--color-text-button-normal-active-6obrzh, #033160)";
export var colorTextButtonNormalDefault = "var(--color-text-button-normal-default-u8647m, #0972d3)";
export var colorTextButtonNormalHover = "var(--color-text-button-normal-hover-kxmvcc, #033160)";
export var colorTextButtonPrimaryActive = "var(--color-text-button-primary-active-vems2x, #ffffff)";
export var colorTextButtonPrimaryDefault = "var(--color-text-button-primary-default-340wh6, #ffffff)";
export var colorTextButtonPrimaryHover = "var(--color-text-button-primary-hover-vzqjpi, #ffffff)";
export var colorTextCounter = "var(--color-text-counter-jwdegc, #5f6b7a)";
export var colorTextDropdownItemFilterMatch = "var(--color-text-dropdown-item-filter-match-x302if, #0972d3)";
export var colorTextDropdownItemHighlighted = "var(--color-text-dropdown-item-highlighted-oaabyk, #000716)";
export var colorTextEmpty = "var(--color-text-empty-fjv325, #5f6b7a)";
export var colorTextFormDefault = "var(--color-text-form-default-yd29ox, #000716)";
export var colorTextFormSecondary = "var(--color-text-form-secondary-ih9x7l, #5f6b7a)";
export var colorTextGroupLabel = "var(--color-text-group-label-w9to2i, #414d5c)";
export var colorTextHeadingDefault = "var(--color-text-heading-default-jopz0f, #000716)";
export var colorTextHeadingSecondary = "var(--color-text-heading-secondary-corkaj, #414d5c)";
export var colorTextHomeHeaderDefault = "var(--color-text-home-header-default-1d71y3, #ffffff)";
export var colorTextHomeHeaderSecondary = "var(--color-text-home-header-secondary-qfyemi, #d1d5db)";
export var colorTextInputDisabled = "var(--color-text-input-disabled-oawyzo, #9ba7b6)";
export var colorTextInputPlaceholder = "var(--color-text-input-placeholder-j53j3c, #5f6b7a)";
export var colorTextInteractiveDefault = "var(--color-text-interactive-default-eg5fsa, #414d5c)";
export var colorTextInteractiveDisabled = "var(--color-text-interactive-disabled-3pbb07, #9ba7b6)";
export var colorTextInteractiveHover = "var(--color-text-interactive-hover-v3lasm, #000716)";
export var colorTextInteractiveInvertedDefault = "var(--color-text-interactive-inverted-default-phki5u, #d1d5db)";
export var colorTextInteractiveInvertedHover = "var(--color-text-interactive-inverted-hover-g2obc6, #fbfbfb)";
export var colorTextLabel = "var(--color-text-label-gxskdy, #000716)";
export var colorTextLinkDefault = "var(--color-text-link-default-5f186r, #0972d3)";
export var colorTextLinkHover = "var(--color-text-link-hover-kololx, #033160)";
export var colorTextNotificationDefault = "var(--color-text-notification-default-2kpcxy, #fbfbfb)";
export var colorTextStatusError = "var(--color-text-status-error-5sesl6, #d91515)";
export var colorTextStatusInactive = "var(--color-text-status-inactive-5ei55p, #5f6b7a)";
export var colorTextStatusInfo = "var(--color-text-status-info-utupcz, #0972d3)";
export var colorTextStatusSuccess = "var(--color-text-status-success-0fzhv9, #037f0c)";
export var colorTextStatusWarning = "var(--color-text-status-warning-27nlyh, #d91515)";
export var fontFamilyBase = "var(--font-family-base-qnistn, 'Open Sans', 'Helvetica Neue', Roboto, Arial, sans-serif)";
export var fontFamilyMonospace = "var(--font-family-monospace-iz223z, Monaco, Menlo, Consolas, 'Courier Prime', Courier, 'Courier New', monospace)";
export var motionEasingResponsive = "var(--motion-easing-responsive-5yj747, cubic-bezier(0, 0, 0, 1))";
export var motionEasingSticky = "var(--motion-easing-sticky-yt83ho, cubic-bezier(1, 0, 0.83, 1))";
export var motionEasingExpressive = "var(--motion-easing-expressive-zns6oe, cubic-bezier(0.84, 0, 0.16, 1))";
export var motionDurationResponsive = "var(--motion-duration-responsive-dvittz, 115ms)";
export var motionDurationExpressive = "var(--motion-duration-expressive-96gzp7, 165ms)";
export var motionDurationComplex = "var(--motion-duration-complex-8zbvzy, 250ms)";
export var motionKeyframesFadeIn = "var(--motion-keyframes-fade-in-9ud20b, awsui-fade-in-35003c)";
export var motionKeyframesFadeOut = "var(--motion-keyframes-fade-out-1ayoyt, awsui-fade-out-35003c)";
export var motionKeyframesStatusIconError = "var(--motion-keyframes-status-icon-error-ti988k, awsui-status-icon-error-35003c)";
export var motionKeyframesScalePopup = "var(--motion-keyframes-scale-popup-3oea16, awsui-scale-popup-35003c)";
export var spaceContainerHorizontal = "var(--space-container-horizontal-ddrezg, 24px)";
export var spaceFieldHorizontal = "var(--space-field-horizontal-89h2yr, 12px)";
export var spaceScaledXxxs = "var(--space-scaled-xxxs-b6dm8t, 2px)";
export var spaceScaledXxs = "var(--space-scaled-xxs-95dhkm, 4px)";
export var spaceScaledXs = "var(--space-scaled-xs-6859qs, 8px)";
export var spaceScaledS = "var(--space-scaled-s-cu1hzn, 12px)";
export var spaceScaledM = "var(--space-scaled-m-pv0fmt, 16px)";
export var spaceScaledL = "var(--space-scaled-l-t03y3z, 20px)";
export var spaceScaledXl = "var(--space-scaled-xl-w2t504, 24px)";
export var spaceScaledXxl = "var(--space-scaled-xxl-pfgv96, 32px)";
export var spaceScaledXxxl = "var(--space-scaled-xxxl-ffokcj, 40px)";
export var spaceStaticXxxs = "var(--space-static-xxxs-k3qmdh, 2px)";
export var spaceStaticXxs = "var(--space-static-xxs-jy4jgd, 4px)";
export var spaceStaticXs = "var(--space-static-xs-9adq92, 8px)";
export var spaceStaticS = "var(--space-static-s-phfs7n, 12px)";
export var spaceStaticM = "var(--space-static-m-pc0uz7, 16px)";
export var spaceStaticL = "var(--space-static-l-t7yjdj, 20px)";
export var spaceStaticXl = "var(--space-static-xl-g2lpmk, 24px)";
export var spaceStaticXxl = "var(--space-static-xxl-m6qwv3, 32px)";
export var spaceStaticXxxl = "var(--space-static-xxxl-66ty0h, 40px)";
