// @cloudscape-design/components/date-picker/use-date-picker.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { useState } from 'react';
import { usePrevious } from '../internal/hooks/use-previous';
import { fireNonCancelableEvent } from '../internal/events';
import { formatDate } from '../internal/utils/date-time';
export function useDatePicker(_a) {
    var value = _a.value, onChange = _a.onChange;
    var defaultSelectedDate = value.length >= 10 ? value : null;
    var _b = useState(defaultSelectedDate), selectedDate = _b[0], setSelectedDate = _b[1];
    var defaultDisplayedDate = value.length >= 10 ? value : formatDate(new Date());
    var _c = useState(defaultDisplayedDate), displayedDate = _c[0], setDisplayedDate = _c[1];
    var onChangeMonthHandler = function (newMonth) {
        setDisplayedDate(formatDate(newMonth));
    };
    var onSelectDateHandler = function (_a) {
        var date = _a.date;
        var formattedDate = formatDate(date);
        setSelectedDate(formattedDate);
        setDisplayedDate(formattedDate);
        fireNonCancelableEvent(onChange, { value: formattedDate });
    };
    var prevValue = usePrevious(value);
    if (prevValue !== value) {
        if (value === '' && selectedDate !== value) {
            setSelectedDate(value);
        }
        // update the displayedDate when inputValue changes in order to
        // display the correct month when the date picker gets open again.
        if (value.length >= 4 && displayedDate !== value) {
            setDisplayedDate(value);
        }
        // set the selected date only when a full date (yyyy-mm-dd) is entered
        if (value.length >= 10 && selectedDate !== value) {
            setSelectedDate(value);
        }
    }
    return {
        defaultDisplayedDate: defaultDisplayedDate,
        displayedDate: displayedDate,
        setDisplayedDate: setDisplayedDate,
        selectedDate: selectedDate,
        setSelectedDate: setSelectedDate,
        onChangeMonthHandler: onChangeMonthHandler,
        onSelectDateHandler: onSelectDateHandler
    };
}
